### ISSUE TITLE
investigate if type erasure can improve compile times

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#10851]
Type erasure is a strategy for compiling generic functions in which the compiled version does is type-independent (the type is "erased"). While it probably can't apply to all Chapel generic functions, perhaps it could apply to some subset of them in order to speed compilation times?

- [ ] write more here
- [ ] discuss inherent non-linear property of generic code with instantiations
- [ ] generic types and methods on them as an example
- [ ] talk through other ideas

**Comment by bradcray** [comment]
Just to make sure I'm following...  We've talked at times about a compiler optimization in which a method on a generic type whose behavior didn't depend on that generic type, such as foo() here:

```chapel
record R {
  param rank: int;
  type eltType;

  proc foo() {
    writeln("Hi from R.foo()!");
  }
}
```

would just result in a single function being generated rather than one per instantiation of `R`.  Is this an instance of what you're proposing or something different?  My sense is "yes", though I'm not seeing how it would result in reduced program performance, making me think I may be wrong.

**Comment by mppf** [comment]
@bradcray - right, Type Erasure *in general* can reduce program performance, but I'm hoping there's an in-between solution that improves compile time without impacting program performance at all (or possibly only impacting it a little).

I think your example above is a good one to think about.

**Comment by bradcray** [comment]
Here's what I was trying to ask in different words:  Do you consider my example to be an instance of type erasure?  If so, would you give an example showing how type erasure can hurt program performance?

**Comment by bradcray** [comment]
OK, thanks.  I was imagining we'd only do this w.r.t. when generic aspects were completely unused.  I understand now that a more aggressive approach might try to combine cases where they were used in which case the method's body would have to deal with that.  For our sake, I'd propose only combining methods for generic aspects that are unused.

**Comment by mppf** [comment]
Right, but it's a bit of a wrinkle. Even in your example, `proc foo` technically "uses" the type eltType, in that there will be a`this` argument and `eltType` will be encoded in it's pointer type. So there's something to do here to figure out how we can handle such things. E.g. right now in the compiler I don't think we can code-generate a pointer to "any instantiation of R".

**Comment by mppf** [comment]
For each generic type, we could consider handling instantiation in 3 ways:

 1. Full instantiation is done by the compiler (what we have now)
 2. The instantiation is unknown until runtime (`any` types in CHIP 2, some similarities to class method dispatch, this is the *type erasure* with impact on runtime performance)
 3. The type consists of two parts - a concrete portion and a generic portion. At runtime, we have only one way to represent the concrete type for all instantiations.

Here 3 represents a new viewpoint (at least to me) that might allow us to make progress towards Brad's goal:

>   I'd propose only combining methods for generic aspects that are unused.

In Brad's example, the instantiated type of R is only used as the type of a `this` argument, but that argument is actually unused. So it could be "erased" and replaced with `void *` or other form of generic pointer.

But what about the common situation where a type has a mix of concrete and generic information?
Additionally what about types like `_array`, which have many methods that exist just to call other methods on a particular type?

``` chapel
record R {
  var x: int;
  type t;
}
proc R.getint() {
  return x;
}
writeln( (new R(1, real)).getint() );
```

Could we store `record R` in two parts, `struct R_concrete { int64_t x; }` and `struct R_instantiated_real { /* type t = real */}` ? Then `struct R_real { R_concrete concrete; R_instantiated_real instantiated; }` ? If we did that, we could represent `R.getint` in C code as
`int64_t R_getint( R_concrete* _this ) { return _this->x; }`. When generating a call to `R.getint` the Chapel compiler would just need to know that it needed to pass the pointer to the concrete fields (rather than the pointer to `R_instantiated_real`). But, with this memory layout, it would always be possible to convert from an `R_instantiated_...` into an `R_concrete` even without looking at the fields. This is similar to how inherited classes work today.

A further extension along these lines would be to think about instantiating each type in stages - so that say setting one generic field makes some fields concrete - but others remain generic. In that way, we we could do something like:

 * DefaultRectangular arrays with different eltType could instantiation details for all methods that do not rely on eltType... e.g. intIdxType, dsiDisplayRepresentation, dsiGetBaseDom, ... these could be instantiated in a way that "type erased" the `eltType` type in particular and so one version could be shared for all DefaultRectangular arrays with the same rank/idxType/etc.

Arguably the above property is already encoded in to DefaultRectangularArr by nature of inheriting from BaseRectangularArr, but the implementation isn't quite ready to do that yet. In particular, it'd be hard for BaseRectangularArr to include the `dom` field, because it's a `DefaultRectangularDom`. However `DefaultRectangularDom` is instantiated with only the type/param information known from the instantiation of `BaseRectangularArr`. So this is a case where the compiler could reduce the number of instantiations. It could consider the parent type in a class inheritance setting to be part of that class' concrete information.

**Comment by mppf** [comment]
I was thinking about this some more recently. One thing I'll note is that today, we can have several types that are trivially different. In particular unmanaged vs borrowed and nilable vs non-nilable class types occupy the same storage. However they are generated as distinct types today.

One thing we could contemplate is to flatten these types after function resolution. That is, say that we have `record R { var x; }` that sometimes stores `unmanaged C` and sometimes stores `borrowed C`. After resolution, we can switch to using one `R` type. Where different instantiated functions need to behave differently, that is already handled because there are no longer generic calls (e.g. one instantiation might include nil checks that another does not).

**Comment by mppf** [comment]
Here is another strategy that might improve function resolution time. The idea is to use a function resolution strategy more similar to constrained generics than the current instantiation strategy. In particular, instead of copying and instantiating a generic function, the generic code stored the types/fn calls to make where relevant? This would allow a later choice to use a runtime binding strategy and would remove unnecessary copies of the same code.

In particular, each generic function would gain an argument bundling up the generic functions that code needs to invoke. This argument is similar to the binding for an interface/constrained generic. The compiler would indicate calls to functions that depend on type using this type-specific behavior argument. It would bundle up the functions needed in a bottom-up manner based on what the generic code invoked. (It is conceptually similar to automatically inferring interfaces).

The idea would be to be able to put off full instantiation until code-generation time. At that time, the instantiation can occur if necessary, for example if information is not re presentable (such as an unknown sized by-value argument or return type).

So generic function resolution would look like this:
 * gather constraints, transitively
 * type check (check gathered constraints are met at call sites)

Then some portion of the compiler would instantiate some of the generics. I hear that the C# CLR takes care to use "monomorphization" only when a vector of generic arguments has a distinct permutation of reference vs value types. I think it's likely that we could come up with a reasonable effective heuristic to balance program performance vs. compilation speed.

Things to be careful of here:
 * reflection
 * overload dispatch

**Comment by bradcray** [comment]
Here something that surprised me today, even though it arguably shouldn't have, related to generic instantiation.  It seems that a potential solution could relate to type erasure, so I'm adding it here:

Recent compilations of Arkouda on master have been complaining about hitting the default instantiation limit (of 256) for `isArray()`.  The point of this instantiation limit, as I recall, is to avoid an infinite instantiation loop and at the time we added the limit (or last considered it?), 256 seemed plenty high.

When I was thinking abstractly about this, I was imagining that perhaps Arkouda had created more than 256 array types which seemed unlikely to me.  It turns out the problem is much simpler.

The implementation of `isArray()` (the version that's hitting the limit) is simply:

```chapel
proc isArray(e)     param  return isArrayValue(e);
```

Looking at isArrayValue(), the definition is similarly fairly simple:

```chapel
  proc isArrayValue(e: []) param  return true;
  proc isArrayValue(e)     param  return false;
```

And the problem is that we call this routine from within module code with all sorts of types:  `_array` records, domain map array classes, various memory management/nilability wrappers to the previous, refs to any of the previous types, iterator records (sooo many iterator records), ranges, domains, etc.  And for each type it's called with, the routine is instantiated.

At first my reaction was naive:  Oh, we should replace calls to `isArray()` in the module code with `isArrayValue()` to avoid creating all these instantiations needlessly.  But then I realized that we'd just be kicking the can down the road by switching from creating 256 instantiations of `isArray()` to 256 instantiations of `isArrayValue()`.  And that was also surprising to me (even though it shouldn't be) because I look at the routines and think "it seems like there should only be two versions of this routine required... one that handles the array case and one that handles the others.  But of course since the signatures of each are generic, we stamp out a copy for each type that's passed in.

So my question here is "Is this something that type erasure could help with?"

I think there are arguably other ways of dealing with this "hitting max instantiations thing":  We could raise the limit (quick fix, but lame), we could mark these core routines as not being subject to the limit (feels neverending), we could detect recursive instantiations more precisely to avoid relying on this cheap way of checking for them (probably reasonable)...  But the thing is we'd still be creating hundreds of instantiations of a function whose purpose is to answer a simple question about a type which feels like a waste of IR.  So if we could come up with a way of creating and reusing a smaller number of copies—particularly for a compile-time function like this that won't be code generated—that seems like a big win (whether related to type erasure or not).

**Comment by mppf** [comment]
@bradcray - I think we could absolutely do better on such functions. One strategy would be to follow through on the comment in https://github.com/chapel-lang/chapel/issues/10851#issuecomment-497364964 - when we bundle up required functions, we would find we don't need any. At that point the compiler could simply rely on the instantiation caching. (In particular, in that design, I'd expect the instantiation caching to go from sets of fns satisfying the requirements to instantiated functions... When there are no fns required it becomes degenerate and there is only one examplar). I personally like this strategy the most of the things discussed in this issue.

A weaker strategy might be to have the compiler observe that a function like `  proc isArrayValue(e)     param  return false;` is so simple that it doesn't even need to do any instantiations. But, generalizing this beyond the very simple case would arrive at the solution I described above.

**Comment by mppf** [comment]
@bradcray - another option (that might feel too much like whack-a-mole) is to change key calls to `isArray` in to calls to `isSubtype` or type comparison with `<=` since these are implemented directly by the compiler and won't require instantiation.

**Comment by bradcray** [comment]
Yeah, that feels like just kicking the problem down the field to me.  For Arkouda's sake, we could also just add "no instantiation limit" pragmas to key routines like this that we run into, but in my notes here, I was really trying to think of ways to solve the general (efficiency) problem, not so much about how to silence Arkouda's issues.  And I think we want users/developers to be able to write patterns like this without undue penalty from a productivity argument (?).

**Comment by vasslitvinov** [comment]
I imagine that for simple "yes/no" functions, it is faster to compute the answer each time, foregoing proper instantiation of the generic FnSymbols. (Not sure how I'd implement this.)

**Comment by mppf** [comment]
As far as https://github.com/chapel-lang/chapel/issues/10851#issuecomment-497364964 goes, one further challenge there is how it combines with type inference. If you have a function containing
``` chapel
proc g(arg) {
  ...
  var x = f(arg);
  ...
  h(x);
}
```

we don't actually know the type of `x` until instantiation (do we?) and as a result it might be challenging to apply the new strategy to such a function, because we couldn't gather the constraints for `h`.

We could make some kind of constraint tree and encode the possible calls to `h` in that, but I'm not sure this is a better representation than the usual AST being instantiated...

But anyway as far as the original question goes, perhaps it would be sufficient to improve compile times if we simply skipped functions that were to complicated in this way.

**Comment by mppf** [comment]
> ``` chapel
> proc g(arg) {
>   ...
>   var x = f(arg);
>   ...
>   h(x);
> }
> ```

I think this case can be reasonably handled by imagining that the compiler adds a new type variable to `g` to indicate the the result type of `f`, like this:
```proc g(arg:?t1, type t2) where f(arg).type == t2```.

I think actually representing such a concept in the compiler would probably use a graph with nodes being generic type variables and edges indicating where one type depends on another; so e.g. `arg` would have an edge to `f(arg)` which represents the type of that function call, and that might in turn have an edge to `x`. Such a graph could be combined with a representation of the functions that need to be available for the generic to compile (so that `x` would have an edge to `h(x)`, and the `f(arg)` and `h(x)` nodes would indicate that these functions needed to be available on those types).
